### PR TITLE
AzureMonitor: Migrate Metrics query editor to React

### LINKED ISSUE #31108: AzureMonitor: Port Metrics UI to React
Related to https://github.com/grafana/grafana/issues/30794

### PR DESCRIPTION
**Pull request by joshhunt**
**What this PR does / why we need it**:

Initial React UI for the query editor, for just the Metrics query type.

image

 - [x] Connect all fields
 - [x] Clear fields when their dependencies change (e.g. when namespace changes, selected metric should be cleared out) 
 - [x] Actually run the query on change
 - [x] Lay out fields properly
 - [x] Hide Angular interface
 - [x] Test dimension filters - I havent been able to get these to populate. Not sure if it's a bug in both the old and new code, or if my data set just doesn't have them.
 
**Which issue(s) this PR fixes**: 

Fixes https://github.com/grafana/grafana/issues/31108

Note, it does not fix the other issues mentioned in the Epic https://github.com/grafana/grafana/issues/30794. They'll have their own PRs :)

### CODE REVIEW

File: public/app/core/angular_wrappers.ts [hunk 1]
```diff
@@ -196,4 +197,11 @@ export function registerAngularDirectives() {
     ['onNowDelayChange', { watchDepth: 'reference', wrapApply: true }],
     ['onHideTimePickerChange', { watchDepth: 'reference', wrapApply: true }],
   ]);
+
+  react2AngularDirective('azureMonitorQueryEditor', QueryEditor, [
+    'query',
+    ['datasource', { watchDepth: 'reference' }],
+    'replaceTemplateVariable',
```
Comment (sunker) [inline public/app/core/angular_wrappers.ts:204]: replaceTemplateVariable is not part of the "real" react [QueryEditorProps](https://github.com/grafana/grafana/blob/master/packages/grafana-data/src/types/datasource.ts#L343-L362) api, so I'd avoid using that. The datasource has access to the template variables, so I suggest you expose a function from the datasource that can expand template variables. This function could then be used from within the query editor components.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  const metricsMetadata = useMetricsMetadata(datasource, query, subscriptionId, replaceTemplateVariable, onQueryChange);
+
+  // Single dynamic onChange function might be a bit unwieldly. Let's see how it goes.
+  // This type magic ensures its only ever called with key/value pairs of the azureMonitor object
+  function onFieldChange<Key extends keyof AzureMonitorQuery['azureMonitor']>(
+    field: Key,
+    { value }: SelectableValue<AzureMonitorQuery['azureMonitor'][Key]>
+  ) {
+    // TODO: investigate empty values. Do they still need to be set?
+    // TODO: when fields change, we actually need to unset all the "lower" fields that depend on this as well
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor.tsx:45]: probably :(

File: public/app/core/angular_wrappers.ts [hunk 3]
```diff
@@ -196,4 +197,11 @@ export function registerAngularDirectives() {
...
+    'onQueryChange',
```
Comment (sunker) [inline public/app/core/angular_wrappers.ts:205]: onChange

File: public/app/core/angular_wrappers.ts [hunk 4]
```diff
@@ -196,4 +197,11 @@ export function registerAngularDirectives() {
...
+  ]);
```
Comment (sunker) [inline public/app/core/angular_wrappers.ts:206]: ...and add onRunQuery. You should be able to run onChange without triggering a panel refresh

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+const MetricsQueryEditor: React.FC<MetricsQueryEditorProps> = ({
+  query,
+  datasource,
+  subscriptionId,
+  replaceTemplateVariable,
+  onQueryChange,
+}) => {
+  const metricsMetadata = useMetricsMetadata(datasource, query, subscriptionId, replaceTemplateVariable, onQueryChange);
+
+  // Single dynamic onChange function might be a bit unwieldly. Let's see how it goes.
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor.tsx:38]: we do the same for Cloud monitoring. Not with this awesome typing though :)

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricNameField.tsx [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        query.azureMonitor.resourceName &&
+        query.azureMonitor.metricNamespace
+      )
+    ) {
+      return;
+    }
+
+    datasource
+      .getMetricNames(
+        replaceTemplateVariable(subscriptionId),
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricNameField.tsx:31]: could we move the expanding of the template variable to the datasource?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricNameField.tsx [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+const MetricName: React.FC<MetricsQueryEditorFieldProps> = ({
+  query,
+  datasource,
+  subscriptionId,
+  replaceTemplateVariable,
+  onChange,
+}) => {
+  const [options, setOptions] = useState<Options>([]);
+  const azureMonitorIsConfigured = datasource.azureMonitorDatasource.isConfigured();
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricNameField.tsx:14]: I know this check wasn't your idea, but it seems to be repeated a lot and I wonder what the point is of checking if the subscription id is set. Isn't it better to just let the query to azure fail if the sub id is not provided?

**Review by sunker — COMMENTED** [review]
Don't have credentials setup for azure at the moment, so couldn't test the look and feel, but the code looks like a great start! Just a few comments

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricNameField.tsx (hunk 7 above)
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/MetricNameField.tsx:14]: Yeah this is fair, ill modify this to do essentially the same thing, but with less duplication.

**Comment by joshhunt** [comment]
whatever, lets just call this ready and review!

Doesn't really contain error handling, that's for a seperate PR (this is gettiung long in the tooth as it is!).

If there's some easy feedback on the UX I'm definitely keen to incorporate that in this PR.

**Comment by joshhunt** [comment]
If @grafana/grafana-frontend-platform has some time, I would appreciate a review from them on this one as well.

I appreciate this PR is a bit large and probably without context, so hit me up if you have any questions :)

**Review by hugohaggmark — COMMENTED** [review]
Probably not the scope for this PR, but we've introduced new support to show standard/custom/ or the same editor as for queries for template variables.

Would be great to have an example for a data source the uses the same Query editor in both Panels/Explore/Variables and we're going to migrate all internal data source plugins to this new API some time soon. 

There are a couple of examples of API usage: TestData, Prometheus and Cloud Monitoring that you can peek at. I'm glad to help answer any questions.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    datasource.azureMonitorDatasource.getSubscriptions().then((results) => {
+      console.log('getSubscriptions results', results);
+      const newSubscriptions = results.map(toOption);
+      setSubscriptions(newSubscriptions);
+
+      // TODO: how much of this is needed?
+      // let newSubscription: string;
+
+      // if (!query.subscription && query.queryType === 'Azure Monitor') {
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx:31]: Good question..I don't know!

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/index.tsx [hunk 9]
```diff
@@ -0,0 +1,99 @@
+import React from 'react';
+
+import Datasource from '../../datasource';
+import { AzureMonitorQuery } from '../../types';
+import { useMetricsMetadata } from '../metrics';
+import SubscriptionField from '../SubscriptionField';
+import MetricNamespaceField from './MetricNamespaceField';
```
[the remark this answers is not in the document]
Comment (sunker) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/index.tsx:7]: Oh so this was already the index file. Never mind!

**Review by sunker — COMMENTED** [review]
Looks great! I'd like to do some manual testing, but still don't have any credentials. Can you help me with that? 

Even though the ux does not have to be perfect at this point, I think we should be able to make some minor improvements. Let's discuss this in the daily. 

Also, are you planning to add unit tests?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx (hunk 8 above)
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/SubscriptionField.tsx:31]: at least thanks for the reminder to look into this more!

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/AggregationField.tsx [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import { findOption } from '../common';
+import { SelectableValue } from '@grafana/data';
+import { AzureQueryEditorFieldProps, Option } from '../../types';
+
+interface AggregationFieldProps extends AzureQueryEditorFieldProps {
+  aggregationOptions: Option[];
+}
+
+const AggregationField: React.FC<AggregationFieldProps> = ({ query, onQueryChange, aggregationOptions }) => {
+  const handleChange = useCallback(
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/AggregationField.tsx:14]: not sure if I'm missing something here, but what's the real benefit of wrapping this function in useCallback?

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts [hunk 11]
```diff
@@ -50,6 +50,8 @@ export default class Datasource extends DataSourceApi<AzureMonitorQuery, AzureDa
   }
 
   query(options: DataQueryRequest<AzureMonitorQuery>): Observable<DataQueryResponse> {
+    console.log('datasource query');
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/datasource.ts:53]: remove this plz

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/types.ts [hunk 12]
```diff
@@ -76,6 +77,7 @@ export interface ApplicationInsightsQuery {
   metricName: string;
   timeGrainUnit: string;
   timeGrain: string;
+  // timeGrainCount: string;
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/types.ts:80]: remove or uncomment

**Review by sunker — CHANGES_REQUESTED** [review]
Great work with the tests! 

One big thing that we've forgotten is the template variables. At the moment they're not displayed as options in the drop downs, and it's not possible to add custom values. You can take a look at how it's done [here](https://github.com/grafana/grafana/blob/master/public/app/plugins/datasource/cloudwatch/components/Stats.tsx#L42). [Here](https://grafana.com/docs/grafana/latest/datasources/azuremonitor/#create-template-variables-for-metrics) are instructions on how to setup template variables for your dashboard. 

Also for the query field, error handling need to be either implemented or the catch statement should be removed.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/response_parser.ts [hunk 13]
```diff
@@ -57,13 +57,17 @@ export default class ResponseParser {
       };
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/azure_monitor/response_parser.ts:57]: yeah, i kind of want to do a rewrite of many of the parts to be honest.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/Field.tsx [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+interface FieldProps extends InlineFieldProps {}
+
+export const Field = (props: FieldProps) => {
+  const { children, ...rest } = props;
+
+  return (
+    <InlineField
+      className={css`
+         {
+          align-items: center;
```
[the remark this answers is not in the document]
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/Field.tsx:17]: no you're right. it was some copy-paste that I just didn't think about.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/AggregationField.tsx (hunk 10 above)
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/MetricsQueryEditor/AggregationField.tsx:14]: useCallback will memoize the function, so it won't 'change' on every render, potentially ca causing unnessisary renders.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/metrics.ts [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+          ...query,
+          azureMonitor: {
+            ...query.azureMonitor,
+            aggregation: metadata.primaryAggType,
+            timeGrain: 'auto',
+            allowedTimeGrainsMs: convertTimeGrainsToMs(metadata.supportedTimeGrains || []),
+          },
+        });
+
+        // TODO: Move the aggregationTypes and timeGrain defaults into `getMetricMetadata`
```
Comment (sunker) [inline public/app/plugins/datasource/grafana-azure-monitor-datasource/components/metrics.ts:59]: is this something that can be done now?

**Review by sunker — COMMENTED** [review]
Looks great! Tested and it works fine. Also tested template variables. 

Just one questions. All pickers have empty catch statements for error handling. What do we want to do there? I don't know how/if this was handled before.

File: public/app/plugins/datasource/grafana-azure-monitor-datasource/components/metrics.ts (hunk 15 above)
Comment (joshhunt) [reply public/app/plugins/datasource/grafana-azure-monitor-datasource/components/metrics.ts:59]: I did it!

**Review by sunker — APPROVED** [review]
LGTM

**Pull request merged**